cachi2/core/package_managers/bundler/parser.py

```diff
@@ -0,0 +1,157 @@
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this git dependency."""
+        return NotImplemented
+
+    @pydantic.field_validator("url")
+    def validate_git_url(cls, url: str) -> str:
+        """Validate that the git URL is a valid git URL."""
+        parsed_url = urlparse(url)
+        if parsed_url.scheme != "https":
+            raise ValueError("GIT dependencies must use HTTPS protocol")
+
+        # TODO: more validation
+
+        return url
+
+    @pydantic.field_validator("revision")
+    def validate_git_revision(cls, revision: str) -> str:
+        """Validate that the git revision is a valid git ref."""
+        if not re.match(r"^[a-fA-F0-9]{40}$", revision):
+            raise ValueError("GIT dependencies must have a valid revision")
+
+        return revision
+
+
+class PathDependency(_GemMetadata):
+    """
+    Represents a path dependency.
+
+    Attributes:
+        path:       Subpath from package root.
+    """
+
+    path: str
+
+    @property
+    def purl(self) -> str:
+        """Get purl for this path dependency."""
+        return NotImplemented
+
+
+BundlerDependency = Union[GemDependency, GitDependency, PathDependency]
+ParseResult = list[BundlerDependency]
+
+
+def parse_lockfile(package_dir: RootedPath) -> ParseResult:
+    """Parse a Gemfile.lock file and return a list of dependencies."""
+    if (
+        not package_dir.join_within_root(GEMFILE).path.exists()
+        or not package_dir.join_within_root(GEMFILE_LOCK).path.exists()
+    ):
+        reason = "Gemfile and Gemfile.lock must be present in the package directory"
+        solution = (
+            "Run `bundle init` to generate the Gemfile.\n"
+            "Run `bundle lock` to generate the Gemfile.lock."
+        )
+        raise PackageRejected(reason=reason, solution=solution)
+
+    bundler_package = files("cachi2.core.package_managers.bundler")
+    lockfile_parser_script = str(bundler_package.joinpath("scripts").joinpath("lockfile_parser.rb"))
+    try:
+        output = run_cmd(cmd=[lockfile_parser_script], params={"cwd": package_dir.path})
+    except subprocess.CalledProcessError:
+        reason = "Execution of Ruby script failed"
+        raise PackageManagerError(reason=reason)
+
+    json_output = json.loads(output)
+
+    bundler_version: str = json_output["bundler_version"]
+    log.info("Package %s is bundled with version %s", package_dir.path.name, bundler_version)
+    dependencies: list[dict[str, str]] = json_output["dependencies"]
+
+    result: ParseResult = []
+    for dep in dependencies:
+        if dep["type"] == "rubygems":
+            result.append(GemDependency.model_validate(dep))
+        elif dep["type"] == "git":
+            result.append(GitDependency.model_validate(dep))
+        elif dep["type"] == "path":
+            _validate_path_dependency_subpath(package_dir, dep["path"])
+            result.append(PathDependency.model_validate(dep))
+
+    return result
```

Comment:
I was doing some checks trying to break the parsing of the `Gemfile.lock` by adding random strings. Apparently, our solution is robust enough to just ignore the random bits :grin:

But I noticed that an empty or pure gibberish `Gemfile.lock` will have a `[]` result, which essentially means that the prefetching for that specific Bundler project will be a noop. 

Do you think we should fail the request in such cases? Or are there any legitimate cases in which the `Gemfile.lock` will have zero dependencies?

Reply:
Zero dependencies is valid (I added a unit test for that as well). But I don't know about completely empty `Gemfile.lock` .

Reply:
From our current parser script output the , there's no way to find out if we have an empty file vs a correctly formed file that has no deps:
```
GEM
  remote: https://rubygems.org/
  specs:

PLATFORMS
  x86_64-linux

DEPENDENCIES

BUNDLED WITH
   2.5.16
```

In the end, this feels like a very cornerish case, so maybe we can just assume this falls under the "outdated lockfile" case.